compiler/modules/types/list.bal

```diff
@@ -572,13 +572,59 @@ function listSubtypeIntersect(Context cx, SubtypeData t1, SubtypeData t2) return
 }
 
 function listSubtypeDiff(Context cx, SubtypeData t1, SubtypeData t2) returns SubtypeData {
+    if listIsEmptySimple(cx, bddIntersect(noBddCache, <Bdd> t1, <Bdd> t2)) {
+        return t1;
+    }
     Bdd b = memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
     if b != false && cx.listMemo[b] == () {
         //io:println("fresh BDD: ", bddToString(b));
     }
     return b;
 }
 
+function listIsEmptySimple(Context cx, Bdd bdd) returns boolean {
+    BddMemo? m = cx.listMemo[bdd];
+    if m !is () && m.empty is boolean {
+        return <boolean>m.empty;
+    }
+    boolean empty = bddEvery(cx, bdd, (), (), isTupleIntersectionEmptySimple);
+    if empty is true {
+        cx.listMemo.put({ bdd, empty }); // handle both memoized values being () and not in memo
+    }
+    return empty;
+}
+
+function isTupleIntersectionEmptySimple(Context cx, Conjunction? pos, Conjunction? neg) returns boolean {
```

Comment:
This name isn't consistent with the rest of this file. This is trying to do listFormulaIsEmpty in a simple way.

This needs to return boolean to work with bddEvery. So I would call it listFormulaIsDefinitelyEmpty. This reflects the semantics: is is true then it definitely is empty; if false, then it's not definitely empty (but may prove to be empty later).

Reply:
fixed